File: java/src/main/java/com/azure/data/cosmos/serialization/hybridrow/recordio/RecordIOParser.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

package com.azure.data.cosmos.serialization.hybridrow.recordio;

import com.azure.data.cosmos.core.Out;
import com.azure.data.cosmos.core.Reference;
import com.azure.data.cosmos.serialization.hybridrow.HybridRowHeader;
import com.azure.data.cosmos.serialization.hybridrow.HybridRowVersion;
import com.azure.data.cosmos.serialization.hybridrow.Result;
import com.azure.data.cosmos.serialization.hybridrow.RowBuffer;
import com.azure.data.cosmos.serialization.hybridrow.SchemaId;
import com.azure.data.cosmos.serialization.hybridrow.io.RowReader;
import com.azure.data.cosmos.serialization.hybridrow.layouts.SystemSchema;
import io.netty.buffer.ByteBuf;
import it.unimi.dsi.fastutil.bytes.Byte2ObjectMap;
import it.unimi.dsi.fastutil.bytes.Byte2ObjectOpenHashMap;
import it.unimi.dsi.fastutil.ints.Int2ObjectMap;
import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;

import javax.annotation.Nonnull;

import static com.google.common.base.Preconditions.checkState;

public final class RecordIOParser {

    private Record record;
    private Segment segment;
    private State state = State.values()[0];

    /**
     * Processes one buffers worth of data possibly advancing the parser state
     *
     * @param buffer1   The buffer to consume
     * @param type     Indicates the type of Hybrid Row produced in {@code record}
     * @param record   If non-empty, then the body of the next record in the sequence
     * @param need     The smallest number of bytes needed to advanced the parser state further
     *                 <p>
     *                 It is recommended that this method not be called again until at least this number of bytes are
     *                 available.
     * @param consumed The number of bytes consumed from the input buffer
     *                 <p>
     *                 This number may be less than the total buffer size if the parser moved to a new state.
     * @return {@link Result#SUCCESS} if no error has occurred;, otherwise the {@link Result} of the last error
     * encountered during parsing.
     * <p>
     * >
     */
    @Nonnull
    public Result process(
        @Nonnull final ByteBuf buffer,
        @Nonnull final Out<ProductionType> type,
        @Nonnull final Out<ByteBuf> record,
        @Nonnull final Out<Integer> need,
        @Nonnull final Out<Integer> consumed) {

        Result result = Result.FAILURE;
        type.set(ProductionType.NONE);
        record.set(null);

        final int start = buffer.readerIndex();

        switch (this.state) {

            case STATE:
                this.state = State.NEED_SEGMENT_LENGTH;
                // TODO: C# TO JAVA CONVERTER: There is no 'goto' in Java:
                //  goto case State.NeedSegmentLength;

            case NEED_SEGMENT_LENGTH: {

                final int minimalSegmentRowSize = HybridRowHeader.BYTES + RecordIOFormatter.SEGMENT_LAYOUT.size();

                if (buffer.readableBytes() < minimalSegmentRowSize) {
                    consumed.set(buffer.readerIndex() - start);
                    need.set(minimalSegmentRowSize);
                    return Result.INSUFFICIENT_BUFFER;
                }

                ByteBuf span = buffer.slice(buffer.readerIndex(), minimalSegmentRowSize);
                RowBuffer row = new RowBuffer(span, HybridRowVersion.V1, SystemSchema.layoutResolver);
                Reference<RowBuffer> tempReference_row =
                    new Reference<RowBuffer>(row);
                RowReader reader = new RowReader(tempReference_row);
                row = tempReference_row.get();
                Reference<RowReader> tempReference_reader =
                    new Reference<RowReader>(reader);
                Out<Segment> tempOut_segment =
                    new Out<Segment>();
                result = SegmentSerializer.Read(tempReference_reader, tempOut_segment);
                this.segment = tempOut_segment.get();
                reader = tempReference_reader.get();
                if (result != Result.SUCCESS) {
                    break;
                }

                this.state = State.NEED_SEGMENT;
                // TODO: C# TO JAVA CONVERTER: There is no 'goto' in Java:
				goto case State.NEED_SEGMENT
            }

            case NEED_SEGMENT: {
                if (buffer.Length < this.segment.length()) {
                    need.set(this.segment.length());
                    consumed.set(buffer.Length - buffer.Length);
                    return Result.INSUFFICIENT_BUFFER;
                }

                //C# TO JAVA CONVERTER WARNING: Unsigned integer types have no direct equivalent in Java:
                //ORIGINAL LINE: Span<byte> span = b.Span.Slice(0, this.segment.Length);
                Span<Byte> span = buffer.Span.Slice(0, this.segment.length());
                RowBuffer row = new RowBuffer(span, HybridRowVersion.V1, SystemSchema.layoutResolver);
                Reference<RowBuffer> tempReference_row2 =
                    new Reference<RowBuffer>(row);
                RowReader reader = new RowReader(tempReference_row2);
                row = tempReference_row2.get();
                Reference<RowReader> tempReference_reader2 =
                    new Reference<RowReader>(reader);
                Out<Segment> tempOut_segment2
                    = new Out<Segment>();
                result = SegmentSerializer.Read(tempReference_reader2, tempOut_segment2);
                this.segment = tempOut_segment2.get();
                reader = tempReference_reader2.get();
                if (result != Result.SUCCESS) {
                    break;
                }

                record.set(buffer.Slice(0, span.Length));
                buffer = buffer.Slice(span.Length);
                need.set(0);
                this.state = State.NEED_HEADER;
                consumed.set(buffer.Length - buffer.Length);
                type.set(ProductionType.SEGMENT);
                return Result.SUCCESS;
            }

            case NEED_HEADER: {
                if (buffer.Length < HybridRowHeader.BYTES) {
                    need.set(HybridRowHeader.BYTES);
                    consumed.set(buffer.Length - buffer.Length);
                    return Result.INSUFFICIENT_BUFFER;
                }

                HybridRowHeader header;
                // TODO: C# TO JAVA CONVERTER: The following method call contained an unresolved 'out' keyword -
                // these cannot be converted using the 'Out' helper class unless the method is within the code
                // being modified:
                MemoryMarshal.TryRead(buffer.Span, out header);
                if (header.Version != HybridRowVersion.V1) {
                    result = Result.INVALID_ROW;
                    break;
                }

                if (SchemaId.opEquals(header.SchemaId,
                    SystemSchema.SEGMENT_SCHEMA_ID)) {
                    // TODO: C# TO JAVA CONVERTER: There is no 'goto' in Java:
					goto case State.NEED_SEGMENT
                }

                if (SchemaId.opEquals(header.SchemaId,
                    SystemSchema.RECORD_SCHEMA_ID)) {
                    // TODO: C# TO JAVA CONVERTER: There is no 'goto' in Java:
					goto case State.NEED_RECORD
                }

                result = Result.INVALID_ROW;
                break;
            }

            case NEED_RECORD: {
                int minimalRecordRowSize = HybridRowHeader.BYTES + RecordIOFormatter.RECORD_LAYOUT.size();
                if (buffer.Length < minimalRecordRowSize) {
                    need.set(minimalRecordRowSize);
                    consumed.set(buffer.Length - buffer.Length);
                    return Result.INSUFFICIENT_BUFFER;
                }

                //C# TO JAVA CONVERTER WARNING: Unsigned integer types have no direct equivalent in Java:
                //ORIGINAL LINE: Span<byte> span = b.Span.Slice(0, minimalRecordRowSize);
                Span<Byte> span = buffer.Span.Slice(0, minimalRecordRowSize);
                RowBuffer row = new RowBuffer(span, HybridRowVersion.V1, SystemSchema.layoutResolver);
                Reference<RowBuffer> tempReference_row3 =
                    new Reference<RowBuffer>(row);
                RowReader reader = new RowReader(tempReference_row3);
                row = tempReference_row3.get();
                Reference<RowReader> tempReference_reader3 = new Reference<RowReader>(reader);
                Out<Record> tempOut_record = new Out<Record>();
                result = RecordSerializer.read(tempReference_reader3, tempOut_record);
                this.record = tempOut_record.get();
                reader = tempReference_reader3.get();
                if (result != Result.SUCCESS) {
                    break;
                }

                buffer = buffer.Slice(span.Length);
                this.state = State.NEED_ROW;
                // TODO: C# TO JAVA CONVERTER: There is no 'goto' in Java:
				goto case State.NEED_ROW
            }

            case NEED_ROW: {
                if (buffer.Length < this.record.length()) {
                    need.set(this.record.length());
                    consumed.set(buffer.Length - buffer.Length);
                    return Result.INSUFFICIENT_BUFFER;
                }

                record.set(buffer.Slice(0, this.record.length()));

                // Validate that the record has not been corrupted.
                //C# TO JAVA CONVERTER WARNING: Unsigned integer types have no direct equivalent in Java:
                //ORIGINAL LINE: uint crc32 = Crc32.Update(0, record.Span);
                int crc32 = Crc32.Update(0, record.get().Span);
                if (crc32 != this.record.crc32()) {
                    result = Result.INVALID_ROW;
                    break;
                }

                buffer = buffer.Slice(this.record.length());
                need.set(0);
                this.state = State.NEED_HEADER;
                consumed.set(buffer.Length - buffer.Length);
                type.set(ProductionType.RECORD);
                return Result.SUCCESS;
            }
        }

        this.state = State.ERROR;
        need.set(0);
        consumed.set(buffer.Length - buffer.Length);
        return result;
    }

    /**
     * True if a valid segment has been parsed.
     */
    public boolean haveSegment() {
        return this.state.value() >= State.NEED_HEADER.value();
    }

    /**
     * If a valid segment has been parsed then current active segment, otherwise undefined.
     */
    public Segment segment() {
        checkState(this.haveSegment());
        return this.segment;
    }

    /**
     * Describes the type of Hybrid Rows produced by the parser.
     */
    public enum ProductionType {
        /**
         * No hybrid row was produced. The parser needs more data.
         */
        NONE(0),

        /**
         * A new segment row was produced.
         */
        SEGMENT(1),

        /**
         * A record in the current segment was produced.
         */
        RECORD(2);

        public static final int BYTES = Integer.BYTES;

        private static Int2ObjectMap<ProductionType> mappings;
        private int value;

        ProductionType(int value) {
            this.value = value;
            mappings().put(value, this);
        }

        public int value() {
            return this.value;
        }

        public static ProductionType from(int value) {
            return mappings().get(value);
        }

        private static Int2ObjectMap<ProductionType> mappings() {
            if (mappings == null) {
                synchronized (ProductionType.class) {
                    if (mappings == null) {
                        mappings = new Int2ObjectOpenHashMap<>();
                    }
                }
            }
            return mappings;
        }
    }

    /**
     * The states for the internal state machine.
     * Note: numerical ordering of these states matters.
     */
    private enum State {
        STATE(
            (byte) 0, "Start: no buffers have yet been provided to the parser"),
        ERROR(
            (byte) 1, "Unrecoverable parse error encountered"),
        NEED_SEGMENT_LENGTH(
            (byte) 2, "Parsing segment header length"),
        NEED_SEGMENT(
            (byte) 3, "Parsing segment header"),
        NEED_HEADER(
            (byte) 4, "Parsing HybridRow header"),
        NEED_RECORD(
            (byte) 5, "Parsing record header"),
        NEED_ROW(
            (byte) 6, "Parsing row body");

        public static final int BYTES = Byte.SIZE;

        private static Byte2ObjectMap<State> mappings;
        private final String description;
        private final byte value;

        State(byte value, String description) {
            this.description = description;
            this.value = value;
            mappings().put(value, this);
        }

        public String description() {
            return this.description;
        }

        public byte value() {
            return this.value;
        }

        public static State from(byte value) {
            return mappings().get(value);
        }

        private static Byte2ObjectMap<State> mappings() {
            if (mappings == null) {
                synchronized (State.class) {
                    if (mappings == null) {
                        mappings = new Byte2ObjectOpenHashMap<>();
                    }
                }
            }
            return mappings;
        }
    }
}

File: src/Serialization/HybridRow/RecordIO/RecordIOParser.cs
// ------------------------------------------------------------
//  Copyright (c) Microsoft Corporation.  All rights reserved.
// ------------------------------------------------------------

namespace Microsoft.Azure.Cosmos.Serialization.HybridRow.RecordIO
{
    using System;
    using System.Runtime.InteropServices;
    using Microsoft.Azure.Cosmos.Core;
    using Microsoft.Azure.Cosmos.Serialization.HybridRow.Schemas;

    public struct RecordIOParser
    {
        private State state;
        private Segment segment;
        private Record record;

        /// <summary>Describes the type of Hybrid Rows produced by the parser.</summary>
        public enum ProductionType
        {
            /// <summary>No hybrid row was produced. The parser needs more data.</summary>
            None = 0,

            /// <summary>A new segment row was produced.</summary>
            Segment,

            /// <summary>A record in the current segment was produced.</summary>
            Record,
        }

        /// <summary>The states for the internal state machine.</summary>
        /// <remarks>Note: numerical ordering of these states matters.</remarks>
        private enum State : byte
        {
            Start = 0, // Start: no buffers have yet been provided to the parser.
            Error, // Unrecoverable parse error encountered.
            NeedSegmentLength, // Parsing segment header length
            NeedSegment, // Parsing segment header
            NeedHeader, // Parsing HybridRow header
            NeedRecord, // Parsing record header
            NeedRow, // Parsing row body
        }

        /// <summary>True if a valid segment has been parsed.</summary>
        public bool HaveSegment => this.state >= State.NeedHeader;

        /// <summary>If a valid segment has been parsed then current active segment, otherwise undefined.</summary>
        public Segment Segment
        {
            get
            {
                Contract.Requires(this.HaveSegment);
                return this.segment;
            }
        }

        /// <summary>Processes one buffers worth of data possibly advancing the parser state.</summary>
        [Obsolete("Use ReadOnlyMemory<byte> override instead.")]
        public Result Process(ReadOnlyMemory<byte> buffer, out ProductionType type, out Memory<byte> record, out int need, out int consumed)
        {
            Result r = this.Process(buffer, out type, out ReadOnlyMemory<byte> rom, out need, out consumed);
            record = MemoryMarshal.AsMemory(rom);
            return r;
        }

        /// <summary>Processes one buffers worth of data possibly advancing the parser state.</summary>
        /// <param name="buffer">The buffer to consume.</param>
        /// <param name="type">Indicates the type of Hybrid Row produced in <paramref name="record" />.</param>
        /// <param name="record">If non-empty, then the body of the next record in the sequence.</param>
        /// <param name="need">
        /// The smallest number of bytes needed to advanced the parser state further. It is
        /// recommended that Process not be called again until at least this number of bytes are available.
        /// </param>
        /// <param name="consumed">
        /// The number of bytes consumed from the input buffer. This number may be less
        /// than the total buffer size if the parser moved to a new state.
        /// </param>
        /// <returns>
        /// <see cref="Microsoft.Azure.Cosmos.Serialization.HybridRow.Result.Success" /> if no error
        /// has occurred, otherwise a valid
        /// <see cref="Microsoft.Azure.Cosmos.Serialization.HybridRow.Result" /> of the last error encountered
        /// during parsing.
        /// </returns>
        /// >
        public Result Process(ReadOnlyMemory<byte> buffer, out ProductionType type, out ReadOnlyMemory<byte> record, out int need, out int consumed)
        {
            Result r = Result.Failure;
            ReadOnlyMemory<byte> b = buffer;
            type = ProductionType.None;
            record = default;
            switch (this.state)
            {
                case State.Start:
                {
                    this.state = State.NeedSegmentLength;
                    goto case State.NeedSegmentLength;
                }

                case State.NeedSegmentLength:
                {
                    int minimalSegmentRowSize = HybridRowHeader.Size + SegmentHybridRowSerializer.Size;
                    if (b.Length < minimalSegmentRowSize)
                    {
                        need = minimalSegmentRowSize;
                        consumed = buffer.Length - b.Length;
                        return Result.InsufficientBuffer;
                    }

                    // TODO: remove this cost-cast when ReadOnlyRowBuffer exists.
                    // The cost-cast implied by MemoryMarshal.AsMemory is only safe here because:
                    // 1. Only READ operations are performed on the row.
                    // 2. The row is not allowed to escape this code.
                    ReadOnlyMemory<byte> mem = b.Slice(0, minimalSegmentRowSize);
                    RowBuffer row = new RowBuffer(MemoryMarshal.AsMemory(mem).Span, 
                        HybridRowVersion.V1, SchemasHrSchema.LayoutResolver);
                    RowCursor root = RowCursor.Create(ref row);
                    r = default(SegmentHybridRowSerializer).Read(ref row, ref root, true, out this.segment);
                    if (r != Result.Success)
                    {
                        break;
                    }

                    this.state = State.NeedSegment;
                    goto case State.NeedSegment;
                }

                case State.NeedSegment:
                {
                    if (b.Length < this.segment.Length)
                    {
                        need = this.segment.Length;
                        consumed = buffer.Length - b.Length;
                        return Result.InsufficientBuffer;
                    }

                    // TODO: remove this cost-cast when ReadOnlyRowBuffer exists.
                    // The cost-cast implied by MemoryMarshal.AsMemory is only safe here because:
                    // 1. Only READ operations are performed on the row.
                    // 2. The row is not allowed to escape this code.
                    ReadOnlyMemory<byte> mem = b.Slice(0, this.segment.Length);
                    RowBuffer row = new RowBuffer(MemoryMarshal.AsMemory(mem).Span, 
                        HybridRowVersion.V1, SchemasHrSchema.LayoutResolver);
                    RowCursor root = RowCursor.Create(ref row);
                    r = default(SegmentHybridRowSerializer).Read(ref row, ref root, true, out this.segment);
                    if (r != Result.Success)
                    {
                        break;
                    }

                    record = b.Slice(0, mem.Length);
                    b = b.Slice(mem.Length);
                    need = 0;
                    this.state = State.NeedHeader;
                    consumed = buffer.Length - b.Length;
                    type = ProductionType.Segment;
                    return Result.Success;
                }

                case State.NeedHeader:
                {
                    if (b.Length < HybridRowHeader.Size)
                    {
                        need = HybridRowHeader.Size;
                        consumed = buffer.Length - b.Length;
                        return Result.InsufficientBuffer;
                    }

                    MemoryMarshal.TryRead(b.Span, out HybridRowHeader header);
                    if (header.Version != HybridRowVersion.V1)
                    {
                        r = Result.InvalidRow;
                        break;
                    }

                    if (header.SchemaId == (SchemaId)SegmentHybridRowSerializer.SchemaId)
                    {
                        goto case State.NeedSegment;
                    }

                    if (header.SchemaId == (SchemaId)RecordHybridRowSerializer.SchemaId)
                    {
                        goto case State.NeedRecord;
                    }

                    r = Result.InvalidRow;
                    break;
                }

                case State.NeedRecord:
                {
                    int minimalRecordRowSize = HybridRowHeader.Size + RecordHybridRowSerializer.Size;
                    if (b.Length < minimalRecordRowSize)
                    {
                        need = minimalRecordRowSize;
                        consumed = buffer.Length - b.Length;
                        return Result.InsufficientBuffer;
                    }

                    // TODO: remove this cost-cast when ReadOnlyRowBuffer exists.
                    // The cost-cast implied by MemoryMarshal.AsMemory is only safe here because:
                    // 1. Only READ operations are performed on the row.
                    // 2. The row is not allowed to escape this code.
                    ReadOnlyMemory<byte> mem = b.Slice(0, minimalRecordRowSize);
                    RowBuffer row = new RowBuffer(MemoryMarshal.AsMemory(mem).Span, 
                        HybridRowVersion.V1, SchemasHrSchema.LayoutResolver);
                    RowCursor root = RowCursor.Create(ref row);
                    r = default(RecordHybridRowSerializer).Read(ref row, ref root, true, out this.record);
                    if (r != Result.Success)
                    {
                        break;
                    }

                    b = b.Slice(mem.Length);
                    this.state = State.NeedRow;
                    goto case State.NeedRow;
                }

                case State.NeedRow:
                {
                    if (b.Length < this.record.Length)
                    {
                        need = this.record.Length;
                        consumed = buffer.Length - b.Length;
                        return Result.InsufficientBuffer;
                    }

                    record = b.Slice(0, this.record.Length);

                    // Validate that the record has not been corrupted.
                    uint crc32 = Crc32.Update(0, record.Span);
                    if (crc32 != this.record.Crc32)
                    {
                        r = Result.InvalidRow;
                        break;
                    }

                    b = b.Slice(this.record.Length);
                    need = 0;
                    this.state = State.NeedHeader;
                    consumed = buffer.Length - b.Length;
                    type = ProductionType.Record;
                    return Result.Success;
                }
            }

            this.state = State.Error;
            need = 0;
            consumed = buffer.Length - b.Length;
            return r;
        }
    }
}
